src/thunderbots/software/geom/util.h

```diff
@@ -0,0 +1,486 @@
+ * P, at which to place a circle to block the cone.
+ *
+ * @pre The cone must have nonzero area.
+ *
+ * @pre \p b must be counterclockwise of \p a.
+ *
+ * @param a the starting angle of the cone.
+ *
+ * @param b the ending angle of the cone.
+ *
+ * @param radius the radius of the circle with which to block the cone.
+ *
+ * @param p the source of the cone.
+ *
+ * @return the blocking position.
+ */
+Point calcBlockCone(const Point &a, const Point &b, const Point &p, const double &radius);
+
+/**
+ * Used for defender_blocks_goal
+ * a = goal post side
+ * c = ball position
+ * g = goalie position
+ * returns the other ray/cone boundary that is not blocked by goalie
+ * I.e. if p is return value,
+ * then points to the other side of line p-c is not covered by goalie.
+ */
+Point calcBlockOtherRay(const Point &a, const Point &c, const Point &g);
+
+/**
+ * Ported code
+ * Checks whether the goalie is in the middle of both goal posts.
+ *
+ * @param a goal post position
+ *
+ * @param b other goal post position
+ *
+ * @param c ball position
+ *
+ * @param g goalie position
+ */
+bool goalieBlockGoalPost(const Point &a, const Point &b, const Point &c, const Point &g);
+
+/**
+ * Calculates a defender position to block the ball.
+ *
+ * @pre the goalie is between the two goal posts, as seen from the ball.
+ *
+ * @param a right goal post position, from goalie's perspective
+ *
+ * @param b other goal post position
+ *
+ * @param c ball position
+ *
+ * @param g goalie position
+ *
+ * @param r radius of defending robot
+ */
+Point calcBlockConeDefender(
+    const Point &a, const Point &b, const Point &c, const Point &g, const double &r);
+
+/*
+ * Ported code from CM geom util
+ */
+
+/**
+ * returns perpendicular offset from line x0-x1 to point p
+ */
+double offsetToLine(Point x0, Point x1, Point p);
+
+/**
+ * returns perpendicular offset from line x0-x1 to point p
+ */
+double offsetAlongLine(Point x0, Point x1, Point p);
+
+/**
+ * returns nearest point on segment a0-a1 to line b0-b1
+ */
+Point segmentNearLine(Point a0, Point a1, Point b0, Point b1);
+
+/**
+ * intersection of two segments?
+ */
+Point intersection(Point a1, Point a2, Point b1, Point b2);
+
+/**
+ * gives counterclockwise angle from <a-b> to <c-b>
+ */
+Angle vertexAngle(Point a, Point b, Point c);
+
+/**
+ * returns time of closest point of approach of two points
+ * moving along constant velocity vectors.
+ */
+double closestPointTime(Point x1, Vector v1, Point x2, Vector v2);
+
+/**
+ * found out if a point is in the vector's direction or against it
+ * if the point normal to the vector, return false
+ *
+ * param[in] offset is the position of the origin of the vector
+ *
+ * param[in] dir is the direction of the vector
+ *
+ * param[in] p is the point is question
+ */
+bool pointInFrontVector(Point offset, Point dir, Point p);
+
+/**
+ * Returns true if v2 is clockwise relative to v1
+ */
+// TODO: this should work but hasn't been properly tested
```

Comment:
-.-

Reply:
This is actually tested, I think the todo is removed now